refactor(ts(x),css): Admin panel component restructure and cleanup

LZoog opened the pull request:
Because:
* Unordered lists and large files were getting hard to maintain. We need to DRY up a lot and make our pages more consistent.

This commit:
* Refactors unordered lists to reusable component tables and rows with X headers or Y headers
* Renames Permissions and AccountSearch to PagePermissions and PageAccountSearch for consistency, moves several components out of Account and into their own component directories
* Creates utils with date format function, shuffles other files around
* Creates component to display 'boolean' values for consistency ona ccount page, e.g. "Yes" or "No"
* Adds HIDE_ROW const
* Layout style tweaks, 'hr' consistency, table visual tweaks
* Removes and adds some Tailwind component classes, installs post-css import
* Shows 'empty string' for redirectURI falsey value

Closes FXA-5253
also closes FXA-4971 (postcss-import in admin-panel)

---

### The diff is different than it appears...

I used `git mv` and did my best to preserve history, however, I think it's because most of these files are more than 50% changed that it's showing some files as deleted and others new. I've spent a couple of hours trying to force git to recognize the file rename to preserve history because this really bothers me, and I haven't had any luck so let me know if you've got any tips after I've already commited and pushed. 😩 I've tried renaming on a separate commit and pulling it in but no matter what, even if `git status` and `git commit --dry-run -a` show "file modified", once I push, it doesn't seem to respect this. Please expand "deleted" files for a better comparison.

We should probably create a follow up to move/make some tests into their new component directories, and potentially increase test coverage. I didn't do that here since we've still technically got coverage and the diff is already so large.

--

<img width="710" alt="image" src="https://user-images.githubusercontent.com/13018240/200018362-736f338e-19b8-41b7-b665-847fa262e4f9.png">

<img width="1146" alt="image" src="https://user-images.githubusercontent.com/13018240/200018550-defd2537-366b-4a48-b172-c430a1cbbef3.png">

<img width="1442" alt="image" src="https://user-images.githubusercontent.com/13018240/200018166-9b59dd1c-c5aa-4fe8-9099-3cb0db1aaed5.png">

![image](https://user-images.githubusercontent.com/13018240/200019298-7721e440-f44c-43ea-92d2-a1e922732c1c.png)

![image](https://user-images.githubusercontent.com/13018240/200021345-de90b29a-41d5-4970-a494-02594c6f628c.png)

![image](https://user-images.githubusercontent.com/13018240/200021423-f1077657-3808-4603-b6e7-72ba7ef28acc.png)

![image](https://user-images.githubusercontent.com/13018240/200021583-ae795f14-101b-4c4e-bd37-0ea6f27398bf.png)

![image](https://user-images.githubusercontent.com/13018240/200021808-b1d8b030-4168-4fe3-abb9-2c1c57af312f.png)

![image](https://user-images.githubusercontent.com/13018240/200019794-4ee86fcb-dcaf-4095-b8d0-4c382208f12e.png)

![image](https://user-images.githubusercontent.com/13018240/200020294-04520c3c-60a0-49d2-a2d6-a2d4e4aadc5c.png)

![image](https://user-images.githubusercontent.com/13018240/200020925-ed5335a3-5c32-45cd-aec7-4246156e4c4f.png)

![image](https://user-images.githubusercontent.com/13018240/200021260-05ae637f-5680-4bde-92e0-d9f9c2ca2396.png)

LZoog commented on packages/fxa-admin-panel/src/components/PageAccountSearch/Account/index.tsx:
```diff
@@ -0,0 +1,389 @@
+  }
+`;
+
+export const LinkedAccount = ({
```
There's even more refactoring we could do here, like `LinkedAccounts` and `AccountDetails` at least could probably use their own components and I'm not sure if we prefer all of our GQL queries/mutations to be in one file but this is a step in the right direction. 🤷‍♀️ Happy to make any further requested changes.

LZoog commented on packages/fxa-admin-panel/src/components/PageAccountSearch/Account/index.tsx:
```diff
@@ -0,0 +1,389 @@
+        )}
+
+        <h3 className="header-lg">Account Recovery Key</h3>
+        {recoveryKeys && recoveryKeys.length > 0 ? (
```
@dschom I didn't dig into the BE but I wonder if we got these confused with recovery codes because I don't know:
* why this is an array? If a user removes their recovery key, it no longer shows up here, so it seems like there should only be one 
* how users can "confirm" their recovery key? Seems like they just create it. I can't seem to "confirm" from Settings. I made this row a Y-header table since with "created at" there's possibly 4 pieces of data to display.

dschom replied:
I just double checked this. The behavior is correct and these are in fact account recovery keys. You are right that there's no reason it needs to be an array, but this doesn't necessarily seem _wrong_ imo.

When recovery keys are created. An email is sent to the user, and then they have to confirm this. So perhaps a better wording would be Email Confirmed.

LZoog replied:
@dschom I just tried it and when I created a new recovery key, I received [this email ](https://storage.googleapis.com/mozilla-storybooks-fxa/commits/latest/fxa-auth-server/index.html?path=/story/fxa-emails-templates-postaddaccountrecovery--post-add-account-recovery) which doesn't ask for confirmation. It looks like the key is just automatically confirmed? 🤔

LZoog commented on packages/fxa-admin-panel/src/components/PageRelyingParties/index.test.tsx:
```diff
@@ -77,7 +77,7 @@ it('renders as expected with a relying party containing all fields', async () =>
   screen.getByText(MOCK_RP_ALL_FIELDS.allowedScopes!);
   screen.getByText('1970', { exact: false });
-  expect(screen.getAllByText('Yes')).toHaveLength(3);
+  expect(screen.getAllByText('true')).toHaveLength(3);
```
I felt like on this page given all the other data we display about RPs, `true` and `false` were nicer for these values. I can easily use the `BooleanResult` here to display "Yes" and "No" consistently if preferred, though.

LZoog commented:
@dschom I've rebased, and after doing so I decided to create another component in `DangerZone` to DRY it up a bit.

Please check me that I got all your Newsletter changes in. I basically had to copy and paste your stuff over for git to be happy since the history here shows deletions.

![image](https://user-images.githubusercontent.com/13018240/200072371-74c72442-a9b6-45c1-a82a-429b03ab14ba.png)

pdehaan commented on packages/fxa-admin-panel/src/components/PageAccountSearch/Account/index.tsx:
```diff
@@ -0,0 +1,390 @@
+      <>{providerId}</>
+      <>{getFormattedDate(authAt)}</>
+      <button
+        className="p-1 text-red-700 border-2 rounded border-grey-100 bg-grey-10 hover:border-2 hover:border-grey-10 hover:bg-grey-50 hover:text-red-700"
```
Does this need `border-2` _and_ `hover:border-2`?

pdehaan commented on packages/fxa-admin-panel/src/components/PageAccountSearch/Account/index.tsx:
```diff
@@ -0,0 +1,390 @@
+
+                <Guard features={[AdminPanelFeature.EditLocale]}>
+                  <button
+                    className="bg-grey-10 border-2 border-grey-100 font-small leading-6 ml-2 rounded text-red-700 w-10 hover:border-2 hover:border-grey-10 hover:bg-grey-50 hover:text-red-700"
```
Possibly duped and unneeded styles:
- `text-red-700` and `hover:text-red-700`
- `border-2` and `hover:border-2`

pdehaan commented on packages/fxa-admin-panel/src/components/PageAccountSearch/Account/index.tsx:
```diff
@@ -0,0 +1,390 @@
+          />
+          <TableRowYHeader
+            header="Created At"
+            value={
+              <>
+                {createdAtDate} ({createdAt})
+              </>
+            }
```
We have this style below, not sure if one is better than the other:
```html
value={`${lockedAtDate} (${lockedAt})`}
```

pdehaan commented on packages/fxa-admin-panel/src/components/PageAccountSearch/Account/index.tsx:
```diff
@@ -0,0 +1,390 @@
+        <h3 className="header-lg">Subscriptions</h3>
+        {subscriptions && subscriptions.length > 0 ? (
+          <>
+            {subscriptions.map((subscription) => (
```
Does this need a type for `subscription`?

dschom replied:
It is not needed in the sense that it is not required, but I think it would be good style to declare it. VSCODE knows it's type, which is MozSubscription.

pdehaan commented on packages/fxa-admin-panel/src/components/PageAccountSearch/ConnectedServices/index.tsx:
```diff
@@ -0,0 +1,162 @@
+        {[
+          location.city || <i>Unknown City</i>,
+          location.state || location.stateCode || '',
+          location.country || location.country || <i>Unknown Country</i>,
```
`location.country || location.country`

dschom replied:
I am guessing it was supposed to be countryCode

LZoog commented on packages/fxa-admin-panel/src/components/TableYHeaders/index.tsx:
```diff
@@ -0,0 +1,59 @@
+}
+
+interface TableRowYHeaderProps {
+  header: string | ReactElement;
```
Hmm, if I take `string` out of the type, I get TS complaints unless I pass in a fragment. A fragment is more flexible, but also adds extra characters and possibly lines since most values are strings. I could change this though.

<img src="https://user-images.githubusercontent.com/13018240/200086447-bc8124fd-a5e5-4032-8382-f6f8fd34d29b.png" width="420px">

![image](https://user-images.githubusercontent.com/13018240/200086573-5ac8cc5d-9a4d-4378-9411-f8b01e85d482.png)

LZoog replied:
@dschom Looking into this a little more, I'd like to actually change this to _just_ `string` for now, at least for the `header` prop. Every instance of this is using a `string` except for one I just changed, so I'd need to change every single instance to a React fragment, and then, that feels a little awkward since `{header}` goes in an `h3` element. I think when we need to use a React element there, that's the time to consider refactoring.

dschom reviewed (approved):
These changes are great 🥇. Thanks for all the effort here! The code will definitely be nicer to work on going forwards.

All my comments definitely land in the nitpick territory. If there's anything that you would like to me file follow up issues on or perhaps simply disagree agree with, please let me know.

LZoog commented on packages/fxa-admin-panel/src/components/TableYHeaders/index.tsx:
```diff
@@ -0,0 +1,59 @@
+
+interface TableRowYHeaderProps {
+  header: string | ReactElement;
+  value?: null | string | ReactElement | ReactElement[];
```
I refactored the [existing `ResultRow value`](https://github.com/mozilla/fxa/blob/f5273536f59b5e93cbd87c3975c316b5c9a903e1/packages/fxa-admin-panel/src/components/AccountSearch/Account/index.tsx#L987) which was typed as `any`. It looks like our connected services can be `null` and that I can change them to pass in `value={os || HIDE_ROW}` for example for now to improve this a bit.

LZoog commented on packages/fxa-admin-panel/src/components/TableYHeaders/index.tsx:
```diff
@@ -0,0 +1,59 @@
+  className,
+}: TableRowYHeaderProps) => {
+  if (!value || value === HIDE_ROW) {
+    return null;
```
Undefined is probably better, this is also what we were doing before.

LZoog commented on packages/fxa-admin-panel/src/components/TableYHeaders/index.tsx:
```diff
@@ -0,0 +1,59 @@
+  return (
+    <tr {...{ className }}>
+      <th className="table-th text-left">{header}</th>
+      <td data-testid={testId} className="table-td border-b">
```
That makes sense, I made `TableXHeaders` after I'd done `TableYHeaders` and can see now how it can probably be more general with the same interface. By "provide a set of elements," do you mean the way `td` is provided?

dschom replied:
I was just thinking a list (i.e. set) of TD elements for each row.

LZoog commented on packages/fxa-admin-panel/src/components/TableXHeaders/index.tsx:
```diff
@@ -0,0 +1,89 @@
+  <>
+    {header && <h3 className="header-lg">{header}</h3>}
+    <table
+      className={`table-x-headers ${borderL && 'border-l-thick'}`}
```
I agree that is definitely more flexible so I'm changing this, but, this also means we have to think a little more about the name of the component class than with this flag since I'm having to send in `className='table-x-headers'` for the ones I've currently got set to `borderL={false}`. I went with what felt more simple but, your suggestion is better in the long run if this continues to grow.

LZoog commented:
@dschom I've addressed all review comments and [filed this follow up ticket](https://mozilla-hub.atlassian.net/browse/FXA-6243) for follow up tweaks, feel free to edit the description of it. I'm still [wondering about the recovery key confirmation](https://github.com/mozilla/fxa/pull/14410#discussion_r1015604936) but I know it's something I can look into myself too. Thanks for the review, I really appreciate all your comments; I'm going to pull the trigger here once CI passes. 😀

dschom commented on packages/fxa-admin-panel/src/components/TableYHeaders/index.tsx:
```diff
@@ -0,0 +1,59 @@
+  testId,
+  className,
+}: TableRowYHeaderProps) => {
+  if (!value || value === HIDE_ROW) {
```
Ah that makes sense now. We gotta keep the backwards compatibility. Now that the table is more general, this definitely seems like a good clean up item.
